// client-side/src/pages/admin/ClassesManagement.jsx
import React, { useEffect, useState, useRef, useCallback, useMemo } from "react";
import { Plus } from 'lucide-react';
import ClassesTable from "../../components/ClassesTable";
import ClassFormDialog from "../../components/ClassFormDialog";
import AssignCoursesDialog from "../../components/AssignCoursesDialog";
import axios from "../../lib/axios";
// UI primitives (custom implementations use default exports only)
import Button from "../../components/ui/Button";
import Label from "../../components/ui/Label";
import Select, { SelectItem } from "../../components/ui/Select";
import { Dialog, DialogContent, DialogHeader, DialogTitle } from "../../components/ui/Dialog";
// pdfmake for client-side PDF generation
import pdfMake from "pdfmake/build/pdfmake.js";
import pdfFonts from "pdfmake/build/vfs_fonts.js";
import { useAuthStore } from "../../store/useAuthStore";
// Courses fetched from server-side endpoint
// Stored in local state for assignment dialog


const initialClassState = {
  name: "",
  term: "",
  year_level: "",
  academic_year: "",
  total_class: 1,
};

export default function ClassesManagement() {
  const authUser = useAuthStore((s) => s.authUser);
  const [classes, setClasses] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState("");
  const [selectedAcademicYear, setSelectedAcademicYear] = useState("all");

  // Dialog states
  const [isAddDialogOpen, setIsAddDialogOpen] = useState(false);
  const [isEditDialogOpen, setIsEditDialogOpen] = useState(false);
  const [isCourseAssignDialogOpen, setIsCourseAssignDialogOpen] = useState(false);
  const [isConfirmDeleteOpen, setIsConfirmDeleteOpen] = useState(false);
  const [classToDelete, setClassToDelete] = useState(null);
  const [deleting, setDeleting] = useState(false);

  // Form states
  const [newClass, setNewClass] = useState(initialClassState);
  const [editingClass, setEditingClass] = useState(null);
  const [assigningClass, setAssigningClass] = useState(null);
  const [selectedCourses, setSelectedCourses] = useState([]);
  const [availableCourses, setAvailableCourses] = useState([]);
  // Selection and export states
  const [selectedIds, setSelectedIds] = useState([]);
  const [isExportDialogOpen, setIsExportDialogOpen] = useState(false);
  const [exportUrl, setExportUrl] = useState("");
  const [exporting, setExporting] = useState(false);
  const [previewMode, setPreviewMode] = useState(false);

  const [page,setPage] = useState(1);
  const [hasMore,setHasMore] = useState(true);
  const limit = 10;
  const loadingRef = useRef(false);
  const sentinelRef = useRef(null);

  const loadClasses = useCallback(async (reset=false)=>{
    if(loadingRef.current) return;
    loadingRef.current = true;
    if(reset){ setPage(1); setHasMore(true); setClasses([]); }
    setLoading(true);
    try {
      const targetPage = reset? 1 : page;
      const res = await axios.get(`/classes?page=${targetPage}&limit=${limit}`);
      const payload = res.data;
      if(Array.isArray(payload)) {
        // legacy non-paginated shape
        setClasses(payload);
        setHasMore(false);
      } else {
        setClasses(prev => reset ? payload.data : [...prev, ...payload.data]);
        setHasMore(payload.hasMore);
      }
      setError("");
    } catch {
      setError("Failed to load classes.");
    } finally {
      setLoading(false);
      loadingRef.current = false;
    }
  },[page,limit]);

  // Fetch classes from API
  useEffect(() => {
    loadClasses(true);
  },[]);
  useEffect(() => {
    if(page>1) loadClasses();
  },[page]);
  useEffect(()=>{
    if(!hasMore) return;
    const el = sentinelRef.current;
    if(!el) return;
    const observer = new IntersectionObserver(entries => {
      const first = entries[0];
      if(first.isIntersecting && hasMore && !loading){
        setPage(p=>p+1);
      }
    }, { threshold: 1 });
    observer.observe(el);
    return ()=> observer.disconnect();
  },[hasMore,loading]);

  // Fetch course catalog
  useEffect(() => {
    axios.get('/courses')
      .then(res => {
        const payload = res.data;
        const list = Array.isArray(payload) ? payload : (Array.isArray(payload.data) ? payload.data : []);
        setAvailableCourses(list);
      })
      .catch(() => console.warn('Failed to load courses list'));
  }, []);

  // Selection helpers for table
  const toggleAll = (checked) => {
    if (!checked) return setSelectedIds([]);
    setSelectedIds(filteredClasses.map(c=>c.id));
  };
  const toggleOne = (id) => {
    setSelectedIds(prev => prev.includes(id) ? prev.filter(x=>x!==id) : [...prev, id]);
  };

  // Build mapping for course details
  const courseMap = useMemo(()=>{
    const m = new Map();
    availableCourses.forEach(c=> m.set(c.course_code, c));
    return m;
  }, [availableCourses]);

  // Generate pdfmake document definition per requirements
  const buildDocDefinition = (classesList) => {
    // Utility: build header blocks for each class
    const content = [];
    const adminDepartment = authUser?.department || 'Admin Department';
    classesList.forEach((cls, idx) => {
      // Header 1
      content.push({
        columns: [
          { text: String(cls.academic_year || ''), bold: true },
          { text: 'CADT', alignment: 'center' },
          { text: 'IDT', alignment: 'center' },
          { text: String(adminDepartment), alignment: 'right' }
        ],
        margin: [0, idx === 0 ? 0 : 20, 0, 2]
      });
      // Header 2
      content.push({
        columns: [
          { text: `Year Level: ${cls.year_level || ''}` },
          { text: `Term: ${cls.term || ''}`, alignment: 'center' },
          { text: `Class: ${cls.name || ''}`, alignment: 'right' }
        ],
        margin: [0, 0, 0, 8]
      });

      // Table body rows from assigned courses
      const rows = [];
      rows.push([
        { text: 'Term', style: 'tableHeader' },
        { text: 'No', style: 'tableHeader' },
        { text: 'Subject', style: 'tableHeader' },
        { text: 'Hour', style: 'tableHeader' },
        { text: 'Credit', style: 'tableHeader' },
        { text: 'Total Class', style: 'tableHeader' },
      ]);
      const courses = Array.isArray(cls.courses) ? cls.courses : [];
      courses.forEach((code, i) => {
        const course = courseMap.get(code) || {};
        rows.push([
          String(cls.term || ''),
          String(i + 1),
          String(course.course_name || code || ''),
          String(course.hours ?? ''),
          String(course.credits ?? ''),
          String(cls.total_class ?? '')
        ]);
      });
      if (!courses.length) {
        rows.push([
          String(cls.term || ''),
          '—',
          'No courses assigned',
          '',
          '',
          String(cls.total_class ?? '')
        ]);
      }
      content.push({
        table: {
          headerRows: 1,
          widths: ['auto','auto','*','auto','auto','auto'],
          body: rows,
        },
        layout: {
          hLineColor: '#e5e7eb',
          vLineColor: '#e5e7eb',
        }
      });
    });

    return {
      pageSize: 'A4',
  pageOrientation: 'landscape',
      pageMargins: [32, 36, 32, 36],
      content,
      styles: {
        tableHeader: { bold: true, fillColor: '#f3f4f6' }
      },
      defaultStyle: { fontSize: 10 }
    };
  };

  // Open preview and build PDF URL
  const openExportPreview = () => {
    setError("");
    setPreviewMode(true);
  };

  const generateSelectedPreview = () => {
    if (!selectedIds.length) {
      setError('Please select at least one class to preview.');
      return;
    }
    const list = filteredClasses.filter(c=> selectedIds.includes(c.id));
    if (!list.length) return;
    setExporting(true);
    try {
      pdfMake.vfs = (pdfFonts && (pdfFonts.vfs || (pdfFonts.pdfMake && pdfFonts.pdfMake.vfs))) || pdfMake.vfs || {};
      const dd = buildDocDefinition(list);
      const pdfDocGenerator = pdfMake.createPdf(dd);
      pdfDocGenerator.getBlob((blob)=>{
        const url = URL.createObjectURL(blob);
        setExportUrl(url);
        setIsExportDialogOpen(true);
      });
    } finally {
      setExporting(false);
    }
  };

  const cancelPreviewSelection = () => {
    setPreviewMode(false);
    setSelectedIds([]);
    setError("");
  };

  const downloadPdf = () => {
    const list = selectedIds.length ? filteredClasses.filter(c=> selectedIds.includes(c.id)) : filteredClasses;
    pdfMake.vfs = (pdfFonts && (pdfFonts.vfs || (pdfFonts.pdfMake && pdfFonts.pdfMake.vfs))) || pdfMake.vfs || {};
    const dd = buildDocDefinition(list);
    pdfMake.createPdf(dd).download(`classes_${Date.now()}.pdf`);
  };


  // Add class
  const handleAddClass = () => {
    setLoading(true);
  const payload = { ...newClass };
  // Normalize total_class (string -> int, default 1)
  const parsedTotal = parseInt(payload.total_class, 10);
  payload.total_class = Number.isFinite(parsedTotal) && parsedTotal > 0 ? parsedTotal : 1;
  // Include any selected courses chosen during add flow
  axios.post("/classes", { ...payload, courses: selectedCourses })
      .then(res => {
        setClasses(prev => [...prev, res.data]);
        setIsAddDialogOpen(false);
        setNewClass(initialClassState);
        setSelectedCourses([]);
        setError("");
      })
      .catch(() => setError("Failed to add class."))
      .finally(() => setLoading(false));
  };

  // Edit class
  const handleEditClass = (classItem) => {
    setEditingClass(classItem);
    setIsEditDialogOpen(true);
  };

  // Update class
  const handleUpdateClass = () => {
    setLoading(true);
  const payload = { ...editingClass };
  const parsedTotal = parseInt(payload.total_class, 10);
  payload.total_class = Number.isFinite(parsedTotal) && parsedTotal > 0 ? parsedTotal : 1;
  axios.put(`/classes/${editingClass.id}`, payload)
      .then(res => {
        setClasses(prev => prev.map(c => c.id === editingClass.id ? res.data : c));
        setIsEditDialogOpen(false);
        setEditingClass(null);
        setError("");
      })
      .catch(() => setError("Failed to update class."))
      .finally(() => setLoading(false));
  };

  // Open confirm delete
  const handleDeleteClass = (classOrId) => {
    const cls = typeof classOrId === 'object' && classOrId !== null
      ? classOrId
      : classes.find(c => String(c.id) === String(classOrId));
    if (!cls) return;
    setClassToDelete(cls);
    setIsConfirmDeleteOpen(true);
  };

  // Perform delete after confirmation
  const performDeleteClass = async () => {
    if (!classToDelete) return;
    setDeleting(true);
    try {
      await axios.delete(`/classes/${classToDelete.id}`);
      setClasses(prev => prev.filter(c => c.id !== classToDelete.id));
      setError("");
      setIsConfirmDeleteOpen(false);
      setClassToDelete(null);
    } catch (e) {
      setError("Failed to delete class.");
    } finally {
      setDeleting(false);
    }
  };

  // Assign courses
  const handleAssignCourses = (classItem) => {
    // If classItem has no id, treat as new-class assignment
    const isNew = !classItem || !classItem.id;
    const target = isNew
      ? { id: null, name: newClass?.name || 'New Class', courses: Array.isArray(selectedCourses) ? selectedCourses : [] }
      : classItem;
    setAssigningClass(target);
    setSelectedCourses(Array.isArray(target.courses) ? target.courses : []);
    axios.get('/courses')
      .then(res => {
        const payload = res.data;
        const list = Array.isArray(payload) ? payload : (Array.isArray(payload.data) ? payload.data : []);
        setAvailableCourses(list);
      })
      .catch((err)=>{ console.debug('Failed to load courses list', err); })
      .finally(()=> setIsCourseAssignDialogOpen(true));
  };

  // Save course assignment
  const handleSaveCourseAssignment = () => {
    // If assigningClass has no id, we're in Add flow: just store selectedCourses and close dialog
    if (!assigningClass || !assigningClass.id) {
      setIsCourseAssignDialogOpen(false);
      setAssigningClass(null);
      return;
    }
    setLoading(true);
    axios.put(`/classes/${assigningClass.id}/courses`, { courses: selectedCourses })
      .then(() => {
        setClasses(prev => prev.map(c => c.id === assigningClass.id ? { ...c, courses: selectedCourses } : c));
        setIsCourseAssignDialogOpen(false);
        setAssigningClass(null);
        setSelectedCourses([]);
        setError("");
      })
      .catch(() => setError("Failed to assign courses."))
      .finally(() => setLoading(false));
  };

  // Toggle course selection
  const handleCourseToggle = (courseCode) => {
    setSelectedCourses(prev => prev.includes(courseCode) ? prev.filter(c => c !== courseCode) : [...prev, courseCode]);
  };

  // Academic year filter
  const getUniqueAcademicYears = () => {
    const years = [...new Set(classes.map(c => c.academic_year))];
    return years.sort();
  };
  const getFilteredClasses = () => {
    if (selectedAcademicYear === "all") return classes;
    return classes.filter(c => c.academic_year === selectedAcademicYear);
  };
  const filteredClasses = getFilteredClasses();

  return (
      <div className="p-4 sm:p-8 space-y-6">
        {/* NOTE: Server-side pagination (limit=10) with infinite scroll for classes implemented. */}
        <div className="flex flex-col gap-3 sm:flex-row sm:items-center sm:justify-between">
          <div>
            <h1 className="text-3xl font-bold text-gray-800 mb-1">Classes Management</h1>
            <p className="text-gray-600">Manage academic classes and assign courses</p>
          </div>
          <Button className="w-full sm:w-auto bg-blue-600 hover:bg-blue-700 text-white flex items-center gap-2" onClick={() => setIsAddDialogOpen(true)}>
            <Plus className="h-4 w-4"/> Add Class
          </Button>
        </div>
        <div className="flex flex-col sm:flex-row sm:items-center gap-3 sm:gap-4">
          <div className="flex w-full sm:w-auto items-center gap-2">
            <Label className="text-gray-700" htmlFor="academic-year-filter">Filter by Academic Year:</Label>
            <div className="w-full sm:w-48">
              <Select
                value={selectedAcademicYear}
                onValueChange={setSelectedAcademicYear}
                placeholder="Select academic year"
              >
                <SelectItem value="all">All Academic Years</SelectItem>
                {getUniqueAcademicYears().map(year => (
                  <SelectItem key={year} value={year}>{year}</SelectItem>
                ))}
              </Select>
            </div>
          </div>
          <div className="text-sm text-gray-500 w-full sm:w-auto">
            Showing <span className="font-medium text-gray-700">{filteredClasses.length}</span> of <span className="font-medium text-gray-700">{classes.length}</span> classes
          </div>
          <div className="w-full sm:w-auto sm:ml-auto flex flex-wrap items-center gap-2">
            {!previewMode ? (
              <Button
                className="w-full sm:w-auto bg-emerald-600 hover:bg-emerald-700 text-white"
                onClick={openExportPreview}
                disabled={!filteredClasses.length}
              >Download</Button>
            ) : (
              <>
                <span className="w-full sm:w-auto text-sm text-gray-600">Selected: <span className="font-medium">{selectedIds.length}</span></span>
                <Button
                  className="w-full sm:w-auto bg-emerald-600 hover:bg-emerald-700 text-white"
                  onClick={generateSelectedPreview}
                  disabled={selectedIds.length === 0}
                >Generate Preview</Button>
                <Button
                  variant="outline"
                  className="w-full sm:w-auto border-gray-200 text-gray-700"
                  onClick={cancelPreviewSelection}
                >Cancel</Button>
              </>
            )}
          </div>
        </div>
        {error && <div className="text-sm mb-2 text-red-600">{error}</div>}
        <div className="overflow-x-auto -mx-4 sm:mx-0">
          <ClassesTable
            classes={filteredClasses}
            onEdit={handleEditClass}
            onDelete={handleDeleteClass}
            onAssignCourses={handleAssignCourses}
            loading={loading}
            courseCatalog={availableCourses}
            title="Academic Classes"
            description="Overview of all academic classes in your department"
            selectable={previewMode}
            selectedIds={selectedIds}
            onToggleOne={toggleOne}
            onToggleAll={toggleAll}
          />
        </div>
        {/* Confirm Delete Dialog */}
        {isConfirmDeleteOpen && classToDelete && (
          <Dialog open={isConfirmDeleteOpen} onOpenChange={(open)=> { setIsConfirmDeleteOpen(open); if (!open) setClassToDelete(null); }}>
            <DialogContent>
              <DialogHeader>
                <DialogTitle>Confirm Deletion</DialogTitle>
              </DialogHeader>
              <div className="px-2 pb-2 text-center space-y-4">
                <p className="text-sm text-gray-700">
                  Do you want to delete this {classToDelete?.name || 'class'}?
                </p>
                <div className="flex flex-col sm:flex-row gap-2 sm:justify-center">
                  <Button
                    onClick={performDeleteClass}
                    className="bg-red-600 hover:bg-red-700 text-white sm:min-w-[120px]"
                    disabled={deleting}
                  >{deleting ? 'Deleting…' : 'OK'}</Button>
                  <Button
                    variant="outline"
                    onClick={()=> { setIsConfirmDeleteOpen(false); setClassToDelete(null); }}
                    className="sm:min-w-[120px]"
                    disabled={deleting}
                  >Cancel</Button>
                </div>
              </div>
            </DialogContent>
          </Dialog>
        )}
        <div ref={sentinelRef} className="h-10 flex items-center justify-center text-xs text-gray-500">
          {loading && hasMore && <span>Loading more...</span>}
          {/* {!hasMore && !loading && <span className="text-gray-400">No more classes</span>} */}
        </div>
        <ClassFormDialog
          open={isAddDialogOpen}
          onOpenChange={setIsAddDialogOpen}
          onSubmit={handleAddClass}
          classData={newClass}
          setClassData={setNewClass}
          isEdit={false}
          onAssignCourses={() => handleAssignCourses(null)}
          selectedCourses={selectedCourses}
          courseCatalog={availableCourses}
        />
  <ClassFormDialog
          open={isEditDialogOpen}
          onOpenChange={setIsEditDialogOpen}
          onSubmit={handleUpdateClass}
          classData={editingClass || initialClassState}
          setClassData={setEditingClass}
          isEdit={true}
        />
        <AssignCoursesDialog
          open={isCourseAssignDialogOpen}
          onOpenChange={setIsCourseAssignDialogOpen}
          availableCourses={availableCourses}
          selectedCourses={selectedCourses}
          onToggleCourse={handleCourseToggle}
          onSave={handleSaveCourseAssignment}
          onCancel={() => setIsCourseAssignDialogOpen(false)}
          className={assigningClass?.name}
        />
        {/* Export Preview Dialog */}
        {isExportDialogOpen && (
          <Dialog open={isExportDialogOpen} onOpenChange={(open)=> { setIsExportDialogOpen(open); if(!open && exportUrl){ URL.revokeObjectURL(exportUrl); setExportUrl(""); } if(!open){ setPreviewMode(false); } }}>
            <DialogContent className="max-w-[1280px] w-[96vw]">
              <DialogHeader>
                <DialogTitle>PDF Preview</DialogTitle>
              </DialogHeader>
              <div className="h-[80vh]">
                {exporting ? (
                  <div className="h-full flex items-center justify-center text-gray-500">Generating preview…</div>
                ) : exportUrl ? (
                  <iframe title="PDF Preview" src={exportUrl} className="w-full h-full border border-gray-200 rounded-md shadow-sm" />
                ) : (
                  <div className="h-full flex items-center justify-center text-gray-400">No preview</div>
                )}
              </div>
              <div className="flex justify-end gap-2">
                <Button variant="outline" onClick={()=> setIsExportDialogOpen(false)}>Close</Button>
                <Button className="bg-emerald-600 hover:bg-emerald-700 text-white" onClick={downloadPdf}>Download</Button>
              </div>
            </DialogContent>
          </Dialog>
        )}
      </div>
  );
}
